src/python_ml_project_template/nets/artflownet.py

```diff
@@ -0,0 +1,214 @@
+import typing
+from dataclasses import dataclass
+from typing import Dict
+
+import plotly.graph_objects as go
+import rpad.pyg.nets.pointnet2 as pnp
+import rpad.visualize_3d.plots as v3p
+import torch
+import torch.optim as opt
+import torch_geometric.data as tgd
+from lightning.pytorch import LightningModule
+from plotly.subplots import make_subplots
+from torch_geometric.data import Batch, Data
+
+
+def flow_metrics(pred_flow, gt_flow):
+    with torch.no_grad():
+        # RMSE
+        rmse = (pred_flow - gt_flow).norm(p=2, dim=1).mean()
+
+        # Cosine similarity, normalized.
+        nonzero_gt_flowixs = torch.where(gt_flow.norm(dim=1) != 0.0)
+        gt_flow_nz = gt_flow[nonzero_gt_flowixs]
+        pred_flow_nz = pred_flow[nonzero_gt_flowixs]
+        cos_dist = torch.cosine_similarity(pred_flow_nz, gt_flow_nz, dim=1).mean()
+
+        # Magnitude
+        mag_error = (pred_flow.norm(p=2, dim=1) - gt_flow.norm(p=2, dim=1)).abs().mean()
+    return rmse, cos_dist, mag_error
+
+
+def artflownet_loss(
+    f_pred: torch.Tensor,
+    f_target: torch.Tensor,
+    n_nodes: torch.Tensor,
+) -> torch.Tensor:
+    """Maniskill loss.
+
+    Args:
+        f_pred (torch.Tensor): Predicted flow.
+        f_target (torch.Tensor): Target flow.
+        n_nodes (torch.Tensor): A list describing the number of nodes
+
+    Returns:
+        Loss
+    """
+
+    # Flow loss, per-point.
+    raw_se = ((f_pred - f_target) ** 2).sum(dim=1)
+
+    # Compute a per-point weighting, so that each point clouQd in a batch
+    # is weighted the same. i.e. so that if one point cloud has 1000 points,
+    # and another point cloud has 15 points, the loss from each point cloud
+    # is equal (by weighting each point's contribution to the loss).
+    weights = (1 / n_nodes).repeat_interleave(n_nodes)
+    l_se = (raw_se * weights).sum()
+
+    # Full loss, aberaged across the batch.
+    loss: torch.Tensor = l_se / len(n_nodes)
+
+    return loss
+
+
+@dataclass
+class ArtFlowNetParams:
+    net: pnp.PN2DenseParams = pnp.PN2DenseParams()
+    mask_input_channel: bool = False
+
+
+class ArtFlowNet(LightningModule):
+    def __init__(
+        self,
+        p: ArtFlowNetParams = ArtFlowNetParams(),
```

Comment:
instead of passing in the parameters, pass in a fully-created network. create that network outside.